DQMServices/Core/src/DQMEDAnalyzer.cc

```diff
@@ -0,0 +1,53 @@
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+#include "DataFormats/Histograms/interface/DQMToken.h"
+
+DQMEDAnalyzer::DQMEDAnalyzer() {
+  produces<DQMToken,edm::Transition::EndLuminosityBlock>();
+  produces<DQMRunToken,edm::Transition::EndRun>();
+}
+
+void DQMEDAnalyzer::beginRun(edm::Run const& run, edm::EventSetup const& setup) 
+{
+  dqmBeginRun(run, setup);
+  edm::Service<DQMStore>()->bookTransaction(
+    [this, &run, &setup](DQMStore::IBooker & booker)
+    {
+      booker.cd();
+      this->bookHistograms(booker, run, setup);
+    },
+    run.run(),
+    run.moduleCallingContext()->moduleDescription()->id());
+}
+
+void DQMEDAnalyzer::endRun(edm::Run const& run, edm::EventSetup const& setup) 
```

Comment:
I would suggest to keep most methods - especially the empty ones - in the .h file.

Reply:
I in general prefer that -- but if it gives me linker errors, I happily put them into a `.cc` file. My theory is that the build system gets more easily confused with implementation in headers (and honestly, I have no idea how this is supposed to work on the linker level, so putting back the `.cc` file was the obvious option).

Reply:
No, the build system does not get "confused".

If you put the implementation of a method inside the class body, e.g.
```c++
class Class {
public:
    void some_method() { ... }
};
```
it is automatically `inline`, and should not create linker problems.

If you put non-inline out-of-body implementation, like
```c++
class Class {
public:
    void some_method();
};

void Class::some_method() { ... }
```
in the .h file, then you do get one version in each .cc file that includes it, and the linker will rightfully compile.

One other thing to pay attention to is if you have virtual methods,  you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

I would be curious what problem you ran into here, and if we can find a less drastic solution than moving all methods' implementations to the .cc file.

Reply:
> One other thing to pay attention to is if you have virtual methods, you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

This one. So my gut feeling of putting things into a .cc file was justifed.

And yes, this inline story works fine as long as the methods are not virtual. Then it gets interesting.